### PR TITLE
aichat: send optional page hostname to feedback api

### PR DESCRIPTION
**Pull request by nullhook**
Resolves https://github.com/brave/brave-browser/issues/35303

- checkbox is enabled by default
- send only http or https
- send only host

image
image

Sec review: https://github.com/brave/reviews/issues/1482

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by fmarier** [comment]
@nullhook Would you mind putting a screenshot in the issue description?

File: components/ai_chat/resources/page/components/feedback_form/index.tsx [hunk 1]
```diff
@@ -86,6 +92,11 @@ function FeedbackForm(props: FeedbackFormProps) {
             />
           </label>
         </fieldset>
+        <fieldset>
+            <Checkbox checked={shouldSendUrl} onChange={handleCheckboxChange}>
+              <label>Share the URL of the page you are on (example.com) to help us make Leo better.</label>
```
Comment (fmarier) [inline components/ai_chat/resources/page/components/feedback_form/index.tsx:97]: Here we should be displaying the real domain that will be sent to our backend instead of `example.com` literally.

Comment (nullhook) [reply components/ai_chat/resources/page/components/feedback_form/index.tsx]: oops, changed and localized the string too

**Review by fmarier — APPROVED** [review]
Thanks for the screenshots.

Looks good to me:
- only HTTP/HTTPS URLs are considered
- there's an explicit checkbox
- the full URL to be collected (currently just the hostname) is displayed in the message

Someone else should do the code review part.

File: components/ai_chat/core/browser/ai_chat_feedback_api.h [hunk 2]
```diff
@@ -36,11 +36,12 @@ class AIChatFeedbackAPI {
...
-                    const std::string& feedback,
-                    const std::string& rating_id,
-                    api_request_helper::APIRequestHelper::ResultCallback
-                        on_complete_callback);
+  void SendFeedback(
+      const std::string& category,
+      const std::string& feedback,
+      const std::string& rating_id,
+      api_request_helper::APIRequestHelper::ResultCallback on_complete_callback,
+      const std::optional<GURL> page_url = std::nullopt);
```
Comment (petemill) [inline components/ai_chat/core/browser/ai_chat_feedback_api.h:44]: nit: feels a little strange to split up the callback and the other params (i.e. to not have the callback at the end of the argument list), no? If you agree, perhaps just don't have the default value and have the caller provide `std::nullopt` when no page url should be sent

File: components/ai_chat/core/browser/conversation_driver.cc [hunk 3]
```diff
@@ -190,7 +191,8 @@ ConversationDriver::GetVisibleConversationHistory() {
   return list;
 }
 
-void ConversationDriver::OnConversationActiveChanged(bool is_conversation_active) {
+void ConversationDriver::OnConversationActiveChanged(
+    bool is_conversation_active) {
```
Comment (petemill) [inline components/ai_chat/core/browser/conversation_driver.cc:195]: looks like there're a few unrelated formatting changes in this PR

File: components/ai_chat/core/browser/ai_chat_feedback_api.h (hunk 2 above)
Comment (nullhook) [reply components/ai_chat/core/browser/ai_chat_feedback_api.h]: makes sense, changed

File: components/ai_chat/core/browser/conversation_driver.cc (hunk 3 above)
Comment (nullhook) [reply components/ai_chat/core/browser/conversation_driver.cc]: fixed

File: components/resources/ai_chat_ui_strings.grdp [hunk 4]
```diff
@@ -247,4 +247,7 @@
   <message name="IDS_CHAT_UI_CONTEXT_TOGGLE_TOOLTIP_INFO" desc="Info tooltip for contextual toggle">
     Toggle on to ask about this page. Its content will be sent to Brave Leo along with your messages.
   </message>
+  <message name="IDS_CHAT_UI_SEND_SITE_URL_LABEL" desc="A checkbox label to opt-in or out for sending current tab's url">
+    Share the URL of the page you are on (<ph name="SITE_URL">$1<ex>www.example.com</ex></ph>) to help us make Leo better.
```
Comment (bbondy) [inline components/resources/ai_chat_ui_strings.grdp:251]: should we change URL to host name here?

Comment (petemill) [reply components/resources/ai_chat_ui_strings.grdp:251]: did you act on this yet @nullhook ?

Comment (nullhook) [reply components/resources/ai_chat_ui_strings.grdp]: changed and updated screenshot

**Review by Brandon-T — APPROVED** [review]

File: components/ai_chat/core/browser/conversation_driver.cc [hunk 5]
```diff
@@ -982,8 +989,10 @@ void ConversationDriver::SendFeedback(
       },
       std::move(callback));
 
-  feedback_api_->SendFeedback(category, feedback, rating_id,
-                              std::move(on_complete));
+  feedback_api_->SendFeedback(
+      category, feedback, rating_id,
+      send_page_url ? std::optional<GURL>(GetPageURL()) : std::nullopt,
```
Comment (petemill) [inline components/ai_chat/core/browser/conversation_driver.cc:994]: I think I would prefer it if we only provided the hostname to the FeedbackAPI here so that we put the restrictive logic all in this class.

Comment (nullhook) [reply components/ai_chat/core/browser/conversation_driver.cc]: ok, i have changed this, `SendFeedback` takes an optional string hostname now.

File: ios/browser/api/ai_chat/ai_chat.mm [hunk 6]
```diff
@@ -189,10 +189,12 @@ - (void)rateMessage:(bool)isLiked
...
             ratingId:(NSString*)ratingId
+         sendPageUrl:(bool)sendPageUrl
           completion:(void (^)(bool))completion {
-  driver_->SendFeedback(
-      base::SysNSStringToUTF8(category), base::SysNSStringToUTF8(feedback),
-      base::SysNSStringToUTF8(ratingId), base::BindOnce(completion));
+  driver_->SendFeedback(base::SysNSStringToUTF8(category),
+                        base::SysNSStringToUTF8(feedback),
+                        base::SysNSStringToUTF8(ratingId), sendPageUrl,
+                        base::BindOnce(completion));
```
Comment (petemill) [inline ios/browser/api/ai_chat/ai_chat.mm:197]: Please rebase and check if iOS fails as a result of this. Might need an additional change to the caller now @Brandon-T

Comment (nullhook) [reply ios/browser/api/ai_chat/ai_chat.mm:197]: rebased

File: components/ai_chat/core/common/mojom/ai_chat.mojom [hunk 7]
```diff
@@ -170,9 +172,7 @@ interface PageHandler {
   // message. |turn_id| is the index of the message in the
   // current conversation.
   RateMessage(bool is_liked, uint32 turn_id) => (string? rating_id);
-  SendFeedback(string category,
-      string feedback,
-          string rating_id) => (bool is_success);
+  SendFeedback(string category, string feedback, string rating_id, bool send_page_url) => (bool is_success);
```
Comment (petemill) [inline components/ai_chat/core/common/mojom/ai_chat.mojom:175]: nit: just noting that half this PR calls this `hostname` (e.g. `send_hostname` or `SiteInfo::hostname`, whereas the other half calls this `url` (e.g. `send_page_url`). That's a little confusing and IMO we should keep it consistent for the same piece of data.

**Review by petemill — APPROVED** [review]

File: components/ai_chat/core/common/mojom/ai_chat.mojom (hunk 7 above)
Comment (nullhook) [reply components/ai_chat/core/common/mojom/ai_chat.mojom]: changed

**Comment by kjozwiak** [comment]
Verification **PASSED** on `Win 11 x64` using the following build(s):

```
Brave | 1.65.2 Chromium: 122.0.6261.43 (Official Build) nightly (64-bit)
-- | --
Revision | 241c70d7f059bd6e2026f12c9928f56f3608e148
OS | Windows 11 Version 23H2 (Build 22631.3155)
```

Using the STR/information outlined via https://github.com/brave/brave-core/pull/21741#issue-2099402708 & https://github.com/brave/brave-browser/issues/35303#issue-2079948345, ensured the following:

* ensured that `Share the hostname of the page you are on` is selected by default
  * also ensured that it's listing the correct `Hostname` within the message (in this example, it's `videocardz.com`)
* ensured that `"domain": videocardz.com"` appears within `https://feedback.brave.com/1/ai/feedback/form`
  * ensured that it's only the `origin` and excludes the `path` and `query params`
* ensured that `domain` doesn't appear within `https://feedback.brave.com/1/ai/feedback/form` when the option is deselected

**`Example`** | **`Example`** | **`Example`** | **`Example`**
----------------|----------------|----------------|-----------------
dislike1 | dislike2 |  dislike3 | dislike4

**Comment by kjozwiak** [comment]
Verification **PASSED** on `Pixel 6` running `Android 14` using the following build(s):

```
Brave | 1.65.4 Chromium: 122.0.6261.43 (Official Build) canary (64-bit)
--- | ---
Revision | 8a887fdf466e5d2de749d90b7a4e898217797665
OS | Android 14; Build/AP11.231215.009; 34; REL
```

Using the STR/information outlined via https://github.com/brave/brave-core/pull/21741#issue-2099402708 & https://github.com/brave/brave-browser/issues/35303#issue-2079948345, ensured the following:

* ensured that `Share the hostname of the page you are on` is selected by default
  * also ensured that it's listing the correct `Hostname` within the message (in this example, it's `videocardz.com`)
* ensured that `"domain": videocardz.com"` appears within `https://feedback.brave.com/1/ai/feedback/form`
  * ensured that it's only the `origin` and excludes the `path` and `query params`
* ensured that `domain` doesn't appear within `https://feedback.brave.com/1/ai/feedback/form` when the option is deselected

**`Example`** | **`Example`** | **`Example`** | **`Example`**
----------------|----------------|----------------|-----------------
Screenshot_20240215-210359 | Screenshot_20240215-210415 | Screenshot 2024-02-15 at 9 04 56 PM | Screenshot_20240215-210420
